experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,1071 @@
+/// would make it small or fast.
+/// See UCPTrieType in ICU4C.
+#[derive(Clone, Copy, PartialEq)]
+pub enum TrieTypeEnum {
+    Any = -1,
+    Fast = 0,
+    Small = 1,
+}
+
+// ValueWidth trait
+
+pub trait ValueWidth: Copy + zerovec::ule::AsULE {
+    // AsUnalignedLittleEndian -> "allowed in a zerovec"
+    const ENUM_VALUE: ValueWidthEnum;
+    fn cast_to_widest(self) -> u32;
+}
+
+impl ValueWidth for u8 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits8;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u16 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits16;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u32 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits32;
+
+    fn cast_to_widest(self) -> u32 {
+        self
+    }
+}
+
+// TrieType trait
+
+pub trait TrieType {
+    const FAST_MAX: u32;
+    const ENUM_VALUE: TrieTypeEnum;
+}
+
+struct Fast;
+impl TrieType for Fast {
+    const FAST_MAX: u32 = 0xffff;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Fast;
+}
+
+struct Small;
+impl TrieType for Small {
+    const FAST_MAX: u32 = 0x0fff;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    index_length: u32,
+    data_length: u32,
+    high_start: u32,
+    shifted12_high_start: u16,
+    index3_null_offset: u16,
+    data_null_offset: u32,
+    null_value: u32,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+pub fn get_code_point_trie_type_enum(trie_type_int: u8) -> TrieTypeEnum {
+    match trie_type_int {
+        0 => TrieTypeEnum::Fast,
+        1 => TrieTypeEnum::Small,
+        _ => TrieTypeEnum::Any,
+    }
+}
+
+pub fn get_code_point_trie_value_width_enum(value_width_int: u8) -> ValueWidthEnum {
+    match value_width_int {
+        0 => ValueWidthEnum::Bits16,
+        1 => ValueWidthEnum::Bits32,
+        2 => ValueWidthEnum::Bits8,
+        _ => ValueWidthEnum::BitsAny,
+    }
+}
+
+// CPTAuto will only expose get_u32() (probably will call it `get()`), but will not / cannot expose get() without using an enum for W
+// CPT<W1,S1> will expose get() and get_u32(). You may not want get_u32() at this point, but it is necessary in order to implement CPTAuto
+
+// u32 vs. char in Rust - unpaired surrogates are valid code points but are not UTF-32 code units, therefore not allowed in `char` type in Rust.
+// But for the CodePointTrie, we do want to be able to look up surrogate code points, therefore we want to use the u32 type. (Note: this is also what we do in the `uniset` crate for `UnicodeSet`.)
+// struct CodePoint(u32) ← enforce whatever invariants you'd like
+
+use impl_const::*;
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    fn trie_internal_small_index(&self, c: u32) -> u32 {
+        let mut i1: u32 = c >> SHIFT_1;
```

Comment:
issue (non-blocking): I know this is a straight port, so this is what the original code looked like, but it would be nice to have descriptive variable names and comments explaining what's going on here (perhaps as a followup)